[TextInput] Add numberOfLines and maximumNumberOfLines props on iOS and fix Android implementation

Szymon20000 opened the pull request:
<!-- Thanks for submitting a pull request! We appreciate you spending the time to work on these changes. Please provide enough information so that others can review your pull request. The three fields below are mandatory. -->

## Summary

We want to be able to set height of TextArea/TextInput/EditText components on all platforms in the same way. Was is currently missing on iOS is expressing the height of a component in lines and this pr is supposed to fix it. Additionally in some cases we want to set maximum number of lines a "TextArea" like component can have. Because of Different screen sizes and accessibility settings it's sometimes hard to compute how many lines a current content takes so the purpose of maximumNumberOfLines prop is to simplify that process.

### iOS How it was implemented?

On Android the native "TextArea" like component (EditText) exposes setLines method that sets height of the component in lines. However, I wasn't able to find anything similar on iOS. What is a common method for both platforms is setting maximum height in lines. On iOS we can set .maximumNumberOfLines property on TextContainer and similarly on Android we can call setMaxLines() method to limit number of lines. So Adding maximumNumberOfLines is not a problem.

I realised that when we add enough number of newLines in order to implement numberOfLines prop using maximumNumberOfLines functionality and did exactly that. Probably it's possible to get a font size and based on that set exact height but I didn't want to miss any edge cases and adding few newlines doesn't seem to be an overhead at all.
Also probably otherwise we would need to take into account things like space between lines, padding...

### Android

It turned out that Android implementation has some issues on Paper and on Fabric the behaviour is shared with TextComponent which limits number of lines instead of setting exact height. So I rewrote the android implementation too. 

### Fabric

I implemented numberOfLines and maximumNumberOfLines similarly to what I did for paper architecture but it turned out that there is one edge case. Whenever you tapped the enter button it added a new line to the text input. Even though NSTextContainer clearly had set maximumNumberOfLines property. I checked what are differences between paper and Fabric implementation when it comes to using NSTextContainer but I wasn't able to spot any difference. Fonts, spaces, other text attributes were exactly the same for every character. In RN docs I found that TextInput already have support for numberOfLines to I checked if the problem occurs also there. Turned out that problem is there too and moreover it's also in paper architecture. I checked differences on how NSContainer is used and spotted a different order of methods called on NSTextContainer, NSTextStorage and NSLayoutManager. After changing the order the problem is gone everywhere.


## Changelog

<!-- Help reviewers and the release process by writing your own changelog entry.

Pick one each for the category and type tags:

[ANDROID|GENERAL|IOS|INTERNAL] [BREAKING|ADDED|CHANGED|DEPRECATED|REMOVED|FIXED|SECURITY] - Message

For more details, see:
https://reactnative.dev/contributing/changelogs-in-pull-requests
-->
[ANDROID][FIXED] - fix numberOfLines for TextInput component on Paper
[ANDROID][CHANGED] - change numberOfLines for TextInput component on Fabric so it's consistent with paper implementation and WEB 
[ANDROID][ADDED] - add maximumNumberOfLines prop to TextInput
[ISO][ADDED] - add numberOfLines for TextInput component on Paper and Fabric
[IOS][ADDED] - add maximumNumberOfLines prop to TextInput (Paper and Fabric)

## Test Plan

<!-- Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots / videos if the pull request changes the user interface. -->

- [ ] test iOS (Paper) 
- [ ] test iOS (Fabric) 
- [ ] test Android (Paper)
- [ ] test Android (Fabric)

necolas commented on Libraries/Components/TextInput/TextInput.js:
```diff
@@ -703,6 +696,17 @@ export type Props = $ReadOnly<{|
+   * Sets the number of lines for a `TextInput`. Use it with multiline set to
+   * `true` to be able to fill the lines.
+   */
+  numberOfLines?: ?number,
```
We intend for this prop to be replaced by the `rows` prop (see above) as per https://github.com/necolas/discussions-and-proposals/blob/reduce-fragmentation/proposals/0000-reduce-fragmentation.md#textinput-component

Szymon20000 replied:
Make sense. Does it mean that I should change the name in this pr or are already working on one yourself?

necolas replied:
The `rows` prop (alias for `numberOfLines`) is already there on L530 ish. It might also make sense to rename (or alias) `maximumNumberOfLines` to something like `maxRows`.

Szymon20000 replied:
Is it android only for now? I haven't seen any logic related to rows on ios.

Szymon20000 replied:
I only see support for maximumNumberOfLines which is not the same. This pr adds support for numberOfLines so if rows hasn't been implemented yet for ios I can rename numberOfLines in the pr to rows.

Szymon20000 replied:
@necolas  are you interested in this pr? Do you want me to rename the prop to rows?

necolas replied:
Yes I already mentioned that. See https://github.com/facebook/react-native/pull/34488

Szymon20000 replied:
@necolas I've just renamed it. Sorry for the delay I was super busy with other tasks.

mateoguzmana commented on ReactCommon/react/renderer/textlayoutmanager/platform/ios/RCTTextLayoutManager.mm:
```diff
@@ -170,26 +170,43 @@ - (LinesMeasurements)getLinesForAttributedString:(facebook::react::AttributedStr
+      paragraphAttributes.maximumNumberOfLines = paragraphAttributes.numberOfLines;
+      NSMutableString *newLines = [NSMutableString stringWithCapacity: paragraphAttributes.numberOfLines];
+      for (NSUInteger i = 0UL; i < paragraphAttributes.numberOfLines; ++i) {
+        // K is added on purpose. New line seems to be not enouth for NTtextContainer
```
```suggestion
        // K is added on purpose. New line seems to be not enough for NTtextContainer
```

Small typo

Szymon20000 replied:
thx :)

janicduplessis commented on Libraries/Text/Text.js:
```diff
@@ -228,6 +228,14 @@ const Text: React.AbstractComponent<
 
   const _hasOnPressOrOnLongPress =
     props.onPress != null || props.onLongPress != null;
+  let numberOfLinesKey = 'numberOfLines';
```
Why do we need to do this? Can we do it in the native implementation of the fabric view instead of having all these conditions in js?

Szymon20000 replied:
In Fabric Most of layout logic and props is shared between TextComponent and TextInputComponent on the native side. I didn't wanted to split them and `numberOfLines` for some reason works differently on Text and TextInput.
For Text is actually sets maximum  height while in TextInput it's exact height. I can probably change it on paper architecture to use maximumNumberOfLines on the native side. Let me know if that would be cleaner. :D

Szymon20000 replied:
Ok, done it this way

janicduplessis commented on Libraries/Text/TextInput/Multiline/RCTMultilineTextInputViewManager.m:
```diff
@@ -17,8 +19,29 @@ - (UIView *)view
+{
+  RCTBaseTextInputShadowView *shadowView = (RCTBaseTextInputShadowView *)[super shadowView];
+ 
+  shadowView.maximumNumberOfLines = 0;
```
Why so we set those to 0 here?

janicduplessis commented on Libraries/Text/TextInput/Multiline/RCTMultilineTextInputViewManager.m:
```diff
@@ -17,8 +19,29 @@ - (UIView *)view
 
 RCT_REMAP_VIEW_PROPERTY(dataDetectorTypes, backedTextInputView.dataDetectorTypes, UIDataDetectorTypes)
 
+RCT_CUSTOM_SHADOW_PROPERTY(maximumNumberOfLines, NSNumber, RCTBaseTextInputShadowView)
```
I think you can use something like this instead https://github.com/facebook/react-native/blob/11570e71a2747602ff485552094b413375b19a96/Libraries/Text/Text/RCTTextViewManager.m#L29

janicduplessis replied:
Probably dont need the remap version for the first one since it has same name

janicduplessis commented:
There are also a few flow errors that will need fixing, you can see on CI.

janicduplessis commented on Libraries/Text/TextInput/RCTBaseTextInputShadowView.m:
```diff
@@ -218,7 +218,16 @@ - (NSAttributedString *)measurableAttributedText
 - (CGSize)sizeThatFitsMinimumSize:(CGSize)minimumSize maximumSize:(CGSize)maximumSize
 {
-  NSAttributedString *attributedText = [self measurableAttributedText];
+  NSMutableAttributedString *attributedText = [[self measurableAttributedText] mutableCopy];
```
I think it might be worth adding a comment that explains that ios doesnt support setting fixed number of lines and what we are doing.

janicduplessis replied:
Same for the fabric part

Szymon20000 replied:
Added! Let me know if the comments make sense.

janicduplessis replied:
Is there a reason for adding the new lines before the text instead of after? Is there any cases where it would affect the width of the input? Also the paper implementation doesn’t need the extra K char to work? I also wonder if there would be a better character to use which has 0 width

Szymon20000 replied:
I think adding after should work well too.

matinzd commented on ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactTextInputManager.java:
```diff
@@ -734,6 +734,11 @@ public void setNumLines(ReactEditText view, int numLines) {
     view.setLines(numLines);
   }
 
+  @ReactProp(name = ViewProps.MAXIMUM_NUMBER_OF_LINES, defaultInt = 1)
```
Isn't it better to have `-1` (UNSET) as default?

Szymon20000 replied:
Yes, I'm rewriting android part now as I found few issues with current implementation. Hope to finish it today.

Szymon20000 commented:
> I think we have differences now between platforms when it comes to how rows/numberOfLines works. (Tested it all in RNTester today)
> 
> ### Android Fabric
> Android Fabric rows actually means max number of rows. So it grows when we add line (as long as we don't go over the limit).
> 
> ### iOS Fabric & Paper
> rows means that height is fixed and expressed in lines no matter what text we display.
> 
> ### WEB (textarea)
> rows means that height is fixed and expressed in lines no matter what text we display. https://developer.mozilla.org/en-US/docs/Web/HTML/Element/textarea#attr-rows
> 
> ### Android Paper
> rows means that height is fixed and expressed in lines no matter what text we display. edit: looks like on paper we set fixed height but it behaves more like minRows :D
> 
> @necolas I think it would make sense to have `rows` and `maxRows`. What do you think?

@matinzd What do you think about it?

Szymon20000 commented:
Now both parts should be fine. 
I will add test to RNTester and test it all one more time on {'iOS', 'Android'} x {'paper', 'fabric'}
One more thing that's left is updating types.

Szymon20000 commented:
@matinzd @necolas Could you guys review my changes? 
I think it's ready :)

necolas commented:
Thanks for working through the feedback so far. We'll need someone from the RN team to review the native code and import the PR

janicduplessis commented on ReactAndroid/src/main/java/com/facebook/react/views/text/TextLayoutManagerMapBuffer.java:
```diff
@@ -405,10 +407,50 @@ public static long measureText(
         maximumNumberOfLines == UNSET || maximumNumberOfLines == 0
             ? layout.getLineCount()
-            : Math.min(maximumNumberOfLines, layout.getLineCount());
+            : Math.min(maximumNumberOfLines, layout.getLineCount());   
```
```suggestion
            : Math.min(maximumNumberOfLines, layout.getLineCount());
```

janicduplessis commented on ReactAndroid/src/main/java/com/facebook/react/views/text/TextLayoutManagerMapBuffer.java:
```diff
@@ -405,10 +407,50 @@ public static long measureText(
+        : UNSET;
+
+    int lines = layout.getLineCount();
+    if (numberOfLines != UNSET && numberOfLines != 0 && numberOfLines > lines && text.length() > 0) {
```
Could we share this logic with TextLayoutManager? Or is this file already kind of just a copy of it?

Szymon20000 replied:
The whole file is a faster copy of TextLayoutManager that uses C++/Java shared buffer instead of using JNI.

janicduplessis commented on ReactCommon/react/renderer/textlayoutmanager/platform/ios/react/renderer/textlayoutmanager/RCTTextLayoutManager.mm:
```diff
@@ -178,26 +178,50 @@ - (LinesMeasurements)getLinesForAttributedString:(AttributedString)attributedStr
+      // K is added on purpose. New line seems to be not enough for NTtextContainer
+      [newLines appendString:@"K\n"];
+    }
+      NSDictionary<NSAttributedStringKey, id> * attributesOfFirstCharacter = [inputAttributedString attributesAtIndex:0 effectiveRange:NULL];
```
```suggestion
    NSDictionary<NSAttributedStringKey, id> * attributesOfFirstCharacter = [inputAttributedString attributesAtIndex:0 effectiveRange:NULL];
```

janicduplessis commented on ReactAndroid/src/main/java/com/facebook/react/views/text/TextLayoutManagerMapBuffer.java:
```diff
@@ -405,10 +407,50 @@ public static long measureText(
+
+      // for some reason a newline on end causes issues with computing height so we add a character
+      if (text.toString().endsWith("\n")) {
+        ssb.append("A");
```
Can we append "K\n" like on iOS to avoid having to do this?

janicduplessis commented on ReactAndroid/src/main/java/com/facebook/react/views/text/TextLayoutManagerMapBuffer.java:
```diff
@@ -405,10 +407,50 @@ public static long measureText(
             ? paragraphAttributes.getInt(PA_KEY_MAX_NUMBER_OF_LINES)
             : UNSET;
 
+      int numberOfLines =
```
```suggestion
    int numberOfLines =
```

janicduplessis reviewed (changes requested):
Can you also add examples in RN Tester of the new props?

Overall looks good, if I understand correctly the changes to Text native component are needed because it is also used by TextInput right?

janicduplessis commented:
Also could you rebase to see if it fixes CI, failures look unrelated.

Szymon20000 commented:
Just added an example in RNTester

Noitidart commented:
This work is so cool, I was just hoping to use this. Thanks so much for this hard work @Szymon20000 and team!

NickGerleman commented on packages/react-native/Libraries/Text/TextInput/RCTBaseTextInputShadowView.m:
```diff
@@ -218,7 +218,22 @@ - (NSAttributedString *)measurableAttributedText
+  NSMutableAttributedString *attributedText = [[self measurableAttributedText] mutableCopy];
+
+  /*
+  * The block below is responsible for setting the exact height of the view in lines
```
Maybe a question for @sammy-SC, is this kosher with any sort of AttributedString layout caching we are doing? I.e. is there something we need to cache bust if layout behavior here varies based on more of the number of line props?

NickGerleman commented on packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/text/ReactBaseTextShadowNode.java:
```diff
@@ -412,6 +413,12 @@ public void setNumberOfLines(int numberOfLines) {
 
+  @ReactProp(name = ViewProps.MAXIMUM_NUMBER_OF_LINES, defaultInt = UNSET)
+  public void setMaxNumberOfLines(int numberOfLines) {
+    mMaxNumberOfLines = numberOfLines == 0 ? UNSET : numberOfLines;
```
Why treat zero as unset?

Szymon20000 replied:
I think it was already the case before with numberOfLines. But seems to have sense. iOS NSTextContainer also uses it as the default value.

NickGerleman commented on packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/text/ReactTextAnchorViewManager.java:
```diff
@@ -49,8 +49,8 @@ public void setAccessible(ReactTextView view, boolean accessible) {
   // maxLines can only be set in master view (block), doesn't really make sense to set in a span
-  @ReactProp(name = ViewProps.NUMBER_OF_LINES, defaultInt = ViewDefaults.NUMBER_OF_LINES)
-  public void setNumberOfLines(ReactTextView view, int numberOfLines) {
+  @ReactProp(name = ViewProps.MAXIMUM_NUMBER_OF_LINES, defaultInt = ViewDefaults.MAXIMUM_NUMBER_OF_LINES)
+  public void setMaxNumberOfLines(ReactTextView view, int numberOfLines) {
```
Hmm, same question as earlier. Are we changing behavior of the existing prop? Seems like this removes the old one in favor of the new one (breaking change).

Szymon20000 replied:
Not really as on the js side it's just mapped. 
On Fabric we already use that new name and I need to make it consistent across arch and platforms.  https://github.com/facebook/react-native/blob/247a913af000aa5912ef9acfca8a834a17c28f59/packages/react-native/ReactCommon/react/renderer/components/text/ParagraphProps.cpp#L81

NickGerleman commented on packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/text/TextLayoutManager.java:
```diff
@@ -389,6 +391,47 @@ public static long measureText(
+    int lines = layout.getLineCount();
+    if (numberOfLines != UNSET && numberOfLines != 0 && numberOfLines >= lines && text.length() > 0) {
+      int numberOfEmptyLines = numberOfLines - lines;
+      SpannableStringBuilder ssb = new SpannableStringBuilder();
```
This is going to be missing all of the spans for things like font size, etc that the user will have set.

Szymon20000 replied:
few lines below I copy all spans

NickGerleman commented on packages/react-native/Libraries/Text/TextInput/RCTBaseTextInputShadowView.m:
```diff
@@ -218,7 +218,22 @@ - (NSAttributedString *)measurableAttributedText
+    for (NSUInteger i = 0UL; i < self.exactNumberOfLines; ++i) {
+      [newLines appendString:@"\n"];
+    }
+    [attributedText insertAttributedString:[[NSAttributedString alloc] initWithString:newLines attributes:self.textAttributes.effectiveTextAttributes] atIndex:0];
```
Does the AttributedString still have all of the various text props that were set?

Text can have things like, different font size in different parts, etc. I'm not sure the most sane way to handle this prop in combination with various metrics.

E.g. we can have an outer TextInput that customizes font size, then a span on the first character which changes it to something else? Which do we pick (it seems like in the java case, we copy all the spans over the beginning of the text).

Szymon20000 replied:
it actually works only when we use same style for the whole TextInput component. I think it makes sense. Let's imagine someone passes Text component as children to textInput and every line has different size. How should that behave? Which line height should we use for missing lines?

NickGerleman commented on packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/text/TextLayoutManager.java:
```diff
@@ -389,6 +391,47 @@ public static long measureText(
+      int numberOfEmptyLines = numberOfLines - lines;
+      SpannableStringBuilder ssb = new SpannableStringBuilder();
+
+      // for some reason a newline on end causes issues with computing height so we add a character
+      if (text.toString().endsWith("\n")) {
+        ssb.append("A");
+      }
```
Could you elaborate on this? Do we see layout issues now when we just have new lines, or does the underlying spannable have some other form we can replicate?

Szymon20000 replied:
I had issues when only new lines were used. There was a weird jump after someone added 2 newlines in a row. It just ignored the second one. This fixed it. It's not visible on the screen as it's only used in layout phase.

NickGerleman commented on packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/text/TextLayoutManager.java:
```diff
@@ -389,6 +391,47 @@ public static long measureText(
+      }
+
+      Object[] spans = text.getSpans(0, 0, Object.class);
+      for (Object span : spans) { // It's possible we need to set exl-exl
```
> It's possible we need to set exl-exl

Should be fine I think, since we never insert to it again?

NickGerleman commented on packages/react-native/ReactCommon/react/renderer/textlayoutmanager/platform/ios/react/renderer/textlayoutmanager/RCTTextLayoutManager.mm:
```diff
@@ -244,26 +244,50 @@ - (void)getRectWithAttributedString:(AttributedString)attributedString
+    NSMutableString *newLines = [NSMutableString stringWithCapacity: paragraphAttributes.numberOfLines];
+    for (NSUInteger i = 0UL; i < paragraphAttributes.numberOfLines; ++i) {
+      // K is added on purpose. New line seems to be not enough for NTtextContainer
+      [newLines appendString:@"K\n"];
```
Same comment as above for Android (but this looks different in the paper case of doing this?).

Szymon20000 replied:
it wasn't working with empty lines for some reason. Similarly to Android.

NickGerleman commented on packages/react-native/ReactCommon/react/renderer/textlayoutmanager/platform/ios/react/renderer/textlayoutmanager/RCTTextLayoutManager.mm:
```diff
@@ -244,26 +244,50 @@ - (void)getRectWithAttributedString:(AttributedString)attributedString
+      // K is added on purpose. New line seems to be not enough for NTtextContainer
+      [newLines appendString:@"K\n"];
+    }
+    NSDictionary<NSAttributedStringKey, id> * attributesOfFirstCharacter = [inputAttributedString attributesAtIndex:0 effectiveRange:NULL];
```
Do we do the same attribute copying in the above copy of this?

Szymon20000 replied:
I don't remember the reason why it's different than on Paper. If that's what you mean. I thin we don't use effectiveAttributes on Fabric? It's just a guess as I don't remember it. Can check if though :)

Szymon20000 commented on packages/react-native/Libraries/Text/Text.js:
```diff
@@ -241,7 +242,7 @@ const Text: React.AbstractComponent<
       isPressable={isPressable}
       nativeID={id ?? nativeID}
-      numberOfLines={numberOfLines}
+      maximumNumberOfLines={numberOfLinesValue}
```
I think so too. Do you think it's better to remove it?

Szymon20000 replied:
Removed!

Szymon20000 commented on packages/react-native/Libraries/Text/Text.js:
```diff
@@ -267,7 +268,7 @@ const Text: React.AbstractComponent<
         isHighlighted={isHighlighted}
         nativeID={id ?? nativeID}
-        numberOfLines={numberOfLines}
+        maximumNumberOfLines={numberOfLinesValue}
```
in Fabric we are already mapping it on the native side as Text component it sets maximumNumberOfLines on textAttributes or paragraphAttributes. There were few native classes shared between TextInputComponent and Text Component and this remapping on the native side made it much cleaner. I can find exactly the example if you want but that seems to be reasonable :) .

NickGerleman replied:
Could we resolve #2 here? We need to ship this as:
1. JS change goes out first
2. Native change goes out a week later

So, we will want to split this change into separate ones for native and JS, where native changes are merged before the JS changes.

In general I would recommend having separate PRs for platforms/sections anyways. I know I personally put quicker to review changes higher up on my priority queue.

Szymon20000 commented on packages/react-native/Libraries/Components/TextInput/AndroidTextInputNativeComponent.js:
```diff
@@ -179,6 +179,13 @@ export type NativeProps = $ReadOnly<{|
+   * `true` to be able to fill the lines.
+   * @platform android
+   */
+  maximumNumberOfLines?: ?Int32,
```
Maybe maxRows? 
I will do my best to apply your suggested changes early next week.

Szymon20000 replied:
I just renamed it in InternalTextInput & flow & TS so that end user can pass `maxNumberOfLines` prop but on the native side we still use `maximumNumberOfLines`. Hope it's fine

Szymon20000 commented on .eslintignore:
```diff
@@ -1,9 +1,9 @@
-packages/react-native/flow/
-packages/react-native/Libraries/Renderer/*
-packages/react-native/Libraries/vendor/**/*
+flow/
+Libraries/Renderer/*
+Libraries/vendor/**/*
```
Probably. Will remove it

Szymon20000 replied:
fixed

Szymon20000 commented on packages/react-native/ReactAndroid/build.gradle:
```diff
@@ -709,6 +709,9 @@ dependencies {
     androidTestImplementation("androidx.test:rules:${ANDROIDX_TEST_VERSION}")
     androidTestImplementation("org.mockito:mockito-core:${MOCKITO_CORE_VERSION}")
 
+    api('com.parse.bolts:bolts-tasks:1.4.0')
+    api('com.parse.bolts:bolts-applinks:1.4.0')
+
```
I think someone broke RN Tester when I was updating that pr and this changes has fixed it. Will remove it!

Szymon20000 commented on packages/react-native/Libraries/Text/TextInput/RCTBaseTextInputShadowView.h:
```diff
@@ -16,6 +16,7 @@ NS_ASSUME_NONNULL_BEGIN
 @property (nonatomic, copy, nullable) NSString *text;
 @property (nonatomic, copy, nullable) NSString *placeholder;
 @property (nonatomic, assign) NSInteger maximumNumberOfLines;
+@property (nonatomic, assign) NSInteger exactNumberOfLines;
```
on fabric yes

Szymon20000 commented on packages/react-native/Libraries/Text/TextInput/Multiline/RCTMultilineTextInputViewManager.m:
```diff
@@ -17,8 +19,21 @@ - (UIView *)view
+{
+  RCTBaseTextInputShadowView *shadowView = (RCTBaseTextInputShadowView *)[super shadowView];
+
+  shadowView.maximumNumberOfLines = 0;
+  shadowView.exactNumberOfLines = 0;
```
I think that values was used on android in shadow nodes so I just copied it. 
It's also used by UIKit https://developer.apple.com/documentation/uikit/nstextcontainer/1444531-maximumnumberoflines

Szymon20000 commented on packages/react-native/Libraries/Text/TextInput/RCTBaseTextInputShadowView.m:
```diff
@@ -218,7 +218,22 @@ - (NSAttributedString *)measurableAttributedText
 - (CGSize)sizeThatFitsMinimumSize:(CGSize)minimumSize maximumSize:(CGSize)maximumSize
 {
-  NSAttributedString *attributedText = [self measurableAttributedText];
+  NSMutableAttributedString *attributedText = [[self measurableAttributedText] mutableCopy];
```
As far as I remember we need to sometimes add characters otherwise the size is not computed properly. 
For instance when we have multiple empty lines. We do it only for layout part and it's not visible on the screen.

Szymon20000 replied:
Ah I think on iOS we only add newlines but still we need to modify the object.

Szymon20000 commented:
@NickGerleman 
I am very grateful that you took your time to check this pr. 🎉
At the beginning of next week I will correct/check the parts you pointed out. 🫡

Szymon20000 commented:
@NickGerleman I applied suggestion that you mentioned and answered all questions. 
Would be great to get another round of review here. 
I try to reply asap.

Szymon20000 commented:
@NickGerleman Would you be able to take a look? 🙂

NickGerleman commented on .eslintignore:
```diff
@@ -1,9 +1,6 @@
 **/main.js
 **/staticBundle.js
 docs/generatedComponentApiDocs.js
-packages/react-native/flow/
-packages/react-native/Libraries/Renderer/*
-packages/react-native/Libraries/vendor/**/*
```
Is this related to your change?

NickGerleman commented:
That mechanism seems like it should work.

I would prefer that we still split this by implementation, and any of the refactoring/renaming, adding conditions to the JS if needed. Right now there is quite a bit of new logic here, along with refactoring which is a little difficult to follow what comes from what.

Szymon20000 commented:
@NickGerleman Just added code in js that checks if maximumNumberOfLines prop is define on the native side if not then it uses the previous name.

NickGerleman commented:
Please split this into different PRs, as per the last comment. This change is too large to review as-is.

Szymon20000 commented:
> Please split this into different PRs, as per the last comment. This change is too large to review as-is.

Sure, Would it be enough to split it into 2 pr android and iOS?

NickGerleman commented:
> > Please split this into different PRs, as per the last comment. This change is too large to review as-is.
> 
> Sure, Would it be enough to split it into 2 pr android and iOS?

I would recommend as granular as possible. iOS/Android is a good split, especially since there are folks on the React Native team with expertise in one, but not the other.

Another split that is common (depending on how much is shared vs not) is splitting between Paper and Fabric.

Szymon20000 commented:
@NickGerleman Here is iOS only pr https://github.com/facebook/react-native/pull/38021
This one will become android only once the iOS is merged. I still need to wrap some C++ parts in if-s as it's shared between android and iOS but the pr can already be reviewed on iOS.

j-piasecki commented on packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactTextInputManager.java:
```diff
@@ -737,9 +737,16 @@ public void setEditable(ReactEditText view, boolean editable) {
 
+  @ReactProp(name = ViewProps.MAXIMUM_NUMBER_OF_LINES, defaultInt = 1)
+  public void setMaxNumLines(ReactEditText view, int numLines) {
+    view.setInputType(view.getInputType() | InputType.TYPE_TEXT_FLAG_MULTI_LINE);
```
```suggestion
    if ((view.getInputType() & InputType.TYPE_TEXT_FLAG_MULTI_LINE) == 0) {
      view.setInputType(view.getInputType() | InputType.TYPE_TEXT_FLAG_MULTI_LINE);
    }
```

We've seen some weird text input flickering on the Expensify app when building for the new arch when the flag is being set on every render.

Szymon20000 replied:
That makes a lot of sense. Thanks for info and quick fix :)

Szymon20000 commented:
This pr turned out to be too complex to review it in one go so I extracted iOS part and moved here ->  
https://github.com/facebook/react-native/pull/38021 
closing the pr so it that it doesn't scare away reviewers :) 
I will open another pr with android fixes once the iOS one is merged.
